x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -842,17 +846,43 @@ protected LogicalPlan rule(LogicalPlan p) {
         }
     }
 
-    /**
-     * A COUNT in a local relation will always be 1.
-     */
-    private static class ReplaceCountInLocalRelation extends OptimizerRule<Aggregate> {
+    static class ReplaceAggregationsInLocalRelations extends OptimizerRule<UnaryPlan> {
 
         @Override
-        protected LogicalPlan rule(Aggregate a) {
-            boolean hasLocalRelation = a.anyMatch(LocalRelation.class::isInstance);
+        protected LogicalPlan rule(UnaryPlan plan) {
+            if (plan instanceof Aggregate || plan instanceof Filter || plan instanceof OrderBy) {
+                LocalRelation relation = unfilteredLocalRelation(plan.child());
+                if (relation != null) {
+                    long count = (relation.executable() instanceof EmptyExecutable ? 0L : 1L);
+
+                    return plan.transformExpressionsDown(AggregateFunction.class, aggregateFunction -> {
+                        if (aggregateFunction instanceof Count) {
+                            return Literal.of(aggregateFunction, count);
+                        } else if (count == 0) {
+                            return Literal.of(aggregateFunction, null);
+                        } else {
+                            // note, most aggregation functions have already been substituted in ReplaceAggregatesWithLiterals
+                            return aggregateFunction;
+                        }
+                    });
+                }
+            }
 
-            return hasLocalRelation ? a.transformExpressionsDown(Count.class, c -> new Literal(c.source(), 1, c.dataType())) : a;
+            return plan;
         }
+
+        private LocalRelation unfilteredLocalRelation(LogicalPlan plan) {
+            List<LogicalPlan> filterOrRelations = plan.collectFirstChildren(p -> p instanceof Filter || p instanceof LocalRelation);
+
+            if (filterOrRelations.size() == 1) {
+                LogicalPlan filterOrRelation = filterOrRelations.get(0);
+                if (filterOrRelation instanceof LocalRelation) {
```

Comment:
Probably on the same idea as @costin's: you collect nodes by the condition on instance of Filter or LocalRelation. After you collect these nodes, you don't do anything in case of Filter, you only do something in case of LocalRelation. Otherwise, you return `null`. So, I don't see the reason for collecting nodes of type Filter.

Looking deeper (at what `collectFirstChildren` is doing), the list of nodes collected can start with a `Filter` or not. If it does start with a filter and it's the first in the list, you ignore it (as being "filtered") and return `null`. If the first children collected is a `LocalRelation` (thus being an "unfiltered" local relation) then you return it. This does raise a question regarding the functionality of the `collectFirstChildren` method which, in this specific case is confusing.

`filterOrRelation` is a misnomer. It's either Relation of `null`. Even more, I'd change this method to return a `LogicalPlan`. If the returned value is `instanceof LocalRelation` then you do something, if it's `instanceof FIlter` then you don't do anything in `rule(UnaryPlan plan)`. I think is clearer like this, rather than returning a `LocalRelation` and comparing to `null` (which in fact is a non-LocalRelation ie a `Filter`).

Reply:
My intention is to get the `LocalRelation` from the first child branch if there is no `Filter` in between. Not sure whether there is a clearer way, but I'm collecting the `Filter` to stop traversal of the branch as soon as a `Filter` has been encountered. Alternatively, I could first check whether a `Filter` exists and then search for the leaf nodes and see whether it's a single `LocalRelation` but this would require two traversals. What would you suggest?

There was a potential issue though when there are multiple leave nodes of which one is a `LocalRelation`. In this case, `filterOrRelations` could have been of size one as well but the transformation is not supported. So I'm collecting all the leaf nodes now.